### PR TITLE
Add email mapper, tests

### PR DESCRIPTION
**Pull request by Noremac201**
1.) Adds support for parsing email oozie action nodes.
2.) Adds tests for EmailMapper
3.) Adds EmailMapper example
4.) Adds new o2a_libs, 'hdfs_helper_functions.py' required by the email mapper.

### CODE REVIEW

**Comment by sprzedwojski** [comment]
Hi! Any chance this and the other pending PRs get merged soon?

**Comment by sprzedwojski** [comment]
@fenglu-g @Noremac201 Hi guys, can we merge the 3 pending pull requests? I'd love to have those changes in `master` and work on top of them :)

**Comment by Noremac201** [comment]
I will leave to @fenglu-g to make sure it's up to his standards/code review for this one

File: oozie-to-airflow/mappers/email_mapper.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+            <body>[BODY]</body>
+            <content_type>[CONTENT-TYPE]</content_type> <!-- content_type is optional -->
+            <attachment>[COMMA-SEPARATED-HDFS-FILE-PATHS]</attachment> <!-- attachment is optional -->
+        </email>
+        <ok to="[NODE-NAME]"/>
+        <error to="[NODE-NAME]"/>
+    </action>
+    """
+
+    def __init__(self, oozie_node, task_id,
```
Comment (sdevani) [inline oozie-to-airflow/mappers/email_mapper.py]: What do we do if there are fields in the node that we don't parse/extract? Can we log it somehow, so that the user is aware of this?

File: oozie-to-airflow/mappers/email_mapper.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+                 trigger_rule=TriggerRule.ALL_SUCCESS, params={}, template='email.tpl'):
+        ActionMapper.__init__(self, oozie_node, task_id, trigger_rule)
+
+        self.template = template
+        self.params = params
+
+        # Parse required nodes first
+        to = xml_utils.find_nodes_by_tag(oozie_node, 'to')[0]
```
Comment (sdevani) [inline oozie-to-airflow/mappers/email_mapper.py]: Nit: 'to' and other strings used should be defined as constants.

File: oozie-to-airflow/o2a_libs/helper_functions.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
+# Copyright 2018 Google LLC
```
Comment (sdevani) [inline oozie-to-airflow/o2a_libs/helper_functions.py]: Can we rename the file to be descriptive of its purpose? Having a generic helper_functions file could result in a dumping grounds in the future. Maybe hdfs_utils.py?

File: oozie-to-airflow/templates/email.tpl [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+    task_id='{{ task_id }}',
+    trigger_rule='{{ trigger_rule }}',
+    params=PARAMS,
+    to={{ to }},
+    subject={{ subject }},
+    html_content={{ body }},
+    cc={{ cc|default(None, true) }},
+    bcc={{ bcc|default(None, true) }},
+    mime_subtype='{{ content_type|default('mixed', true) }}',
+    files=read_hdfs_files({{ attachment|default([], true) }})
```
Comment (sdevani) [inline oozie-to-airflow/templates/email.tpl]: I'm new to this repo, so excuse my silly questions.

Does this mean that the output DAG will be dependent on this oozie-to-airflow lib? That sounds like it makes intuitive sense IMO.

File: oozie-to-airflow/tests/mappers/test_email_mapper.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        self.assertEqual(['/file/in/hdfs.txt'], mapper.attachment)
+
+    def test_convert_to_text(self):
+        self._create_xml_no_jinja()
+        mapper = email_mapper.EmailMapper(oozie_node=self.et.getroot(),
+                                          task_id='test_id',
+                                          trigger_rule=TriggerRule.DUMMY)
+        # Throws a syntax error if doesn't parse correctly
+        ast.parse(mapper.convert_to_text())
+
```
Comment (sdevani) [inline oozie-to-airflow/tests/mappers/test_email_mapper.py]: Can we add a test case for defaults when fields are not specified. Maybe 1 test case for regular defaults and 1 test case for when a critical field is missing.

File: oozie-to-airflow/tests/o2a_libs/test_helper_functions.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        OUTPUT = 'data'
+        check_mock.return_value = OUTPUT.encode('utf-8')
+
+        with mock.patch("builtins.open", mock.mock_open(read_data="data")) as mock_file:
+            for f in helper_functions.read_hdfs_files(['test.txt']):
+                self.assertEqual('data', f.read())
+                check_mock.assert_called_once_with(['hadoop', 'fs', '-cat', 'test.txt'])
+
+    @mock.patch('subprocess.check_output')
+    def test_read_hdfs_file(self, check_mock):
```
Comment (sdevani) [inline oozie-to-airflow/tests/o2a_libs/test_helper_functions.py]: Can we test for cases where the subprocess fails?

File: oozie-to-airflow/tests/o2a_libs/test_helper_functions.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+from o2a_libs import helper_functions
+
+
+class TestHelperFunctions(unittest.TestCase):
+
+    def test_read_hdfs_files_empty(self):
+        self.assertEqual([], helper_functions.read_hdfs_files([]))
+
+    @mock.patch('subprocess.check_output')
+    def test_read_hdfs_files(self, check_mock):
```
Comment (sdevani) [inline oozie-to-airflow/tests/o2a_libs/test_helper_functions.py]: Test cases for empty array and for multiple files.

File: oozie-to-airflow/utils/el_utils.py [hunk 8]
```diff
@@ -93,15 +94,19 @@ def parse_el_func(el, el_func_map=EL_FUNCTIONS):
...
-        raise KeyError('{} EL function not supported.'.format(fn_match[0][0]))
+    parsed_fns = []
+    for fn in fn_match:
+        # fn_match is of the form [('concat', "'ls', '-l'")]
+        # for an el function like ${concat('ls', '-l')}
+        if fn[0] not in el_func_map:
+            raise KeyError('{} EL function not supported.'.format(fn[0]))
 
-    mapped_func = el_func_map[fn_match[0][0]]
+        mapped_func = el_func_map[fn_match[0][0]]
```
Comment (sdevani) [inline oozie-to-airflow/utils/el_utils.py]: I might not be understanding this correctly, but it looks like we're iterating through fn_match, but each time we're only using fn_match[0][0]. When not fn[0] instead?

At the surface level it seems like this can potentially yield incorrect behavior.

File: oozie-to-airflow/utils/el_utils.py [hunk 9]
```diff
@@ -11,13 +11,16 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
```
Comment (sdevani) [inline oozie-to-airflow/utils/el_utils.py]: The changes to el_utils seem independent of the changes to email mapper. Can we split this into separate commits and likely separate PRs?

File: oozie-to-airflow/mappers/email_mapper.py [hunk 10]
```diff
@@ -12,14 +12,16 @@
...
 # limitations under the License.
 import os
-import subprocess
-
 import jinja2
+
+from airflow.operators import email_operator
 from airflow.utils.trigger_rule import TriggerRule
 
-from definintions import ROOT_DIR
```
Comment (sdevani) [inline oozie-to-airflow/mappers/email_mapper.py]: This should be fixed in the prior commit. Each commit should be good to go stand-alone.

File: oozie-to-airflow/mappers/email_mapper.py [hunk 11]
```diff
@@ -12,14 +12,16 @@
 # See the License for the specific language governing permissions and
```
Comment (sdevani) [inline oozie-to-airflow/mappers/email_mapper.py]: Can we add tests for this new method?

File: oozie-to-airflow/utils/el_utils.py [hunk 12]
```diff
@@ -143,7 +143,8 @@ def convert_el_to_jinja(oozie_el, quote=True):
         parsed_fns = parse_el_func(jinjafied_el, EL_FUNCTIONS)
         return template.render(str=subbed_str, els=parsed_fns)
 
-    return '\'' + jinjafied_el + '\'' if quote else jinjafied_el
+    # This takes care of any escaping that must be done for python to understand
+    return repr(jinjafied_el) if quote else repr(jinjafied_el)[1:-1]
```
Comment (sdevani) [inline oozie-to-airflow/utils/el_utils.py]: Can we add some test cases. This change should be independent of the doc update.

File: oozie-to-airflow/init.sh [hunk 13]
```diff
@@ -17,7 +17,7 @@
 command -v pip >/dev/null 2>&1 || { echo >&2 "pip appears to not be installed. Aborting."; exit 1; }
 
 # Check for python3
-command -v python3 >/dev/null 2>&1 || { echo >&2 "python 3 does not appear to be installed. Aborting."; exit 1;}
+command -v python3 >/dev/null 2>&1 || { echo >&2 "python 3 does not appear to be installed. Aborting."; exit 1; }
```
Comment (sdevani) [inline oozie-to-airflow/init.sh]: I haven't used git in a while so feel free to correct me.

Is this a result of merging in changes from upstream into your working branch? If so, would rebasing help prevent his?

File: oozie-to-airflow/templates/email.tpl (hunk 4 above)
Comment (Noremac201) [reply oozie-to-airflow/templates/email.tpl]: In a sense, but only the `o2a_libs/` folder. The way it's currently set up is that the o2a_libs/ directory must be present in the user's DAG folder. This will allow airflow to pick it up and use it on all the converted DAGs.

File: oozie-to-airflow/mappers/email_mapper.py [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+        if content_type:
+            # Don't quote this since the quote is already in the template (required by jinja default option)
+            self.content_type = el_utils.replace_el_with_var(content_type[0].text, params=self.params, quote=False)
+
+        attachment = xml_utils.find_nodes_by_tag(oozie_node, 'attachment')
+        if attachment:
+            self.attachment = el_utils.replace_el_with_var(attachment[0].text, params=self.params, quote=False)
+            self.attachment = self.attachment.split(",")
+
+    def convert_to_text(self):
```
[the remark this answers is not in the document]
Comment (Noremac201) [reply oozie-to-airflow/mappers/email_mapper.py]: I am unaware the direction we are taking -- I see two possible choices.
The one you said, throw an error, or the other choice, which is to log it, and do a best effort conversion.

Let me know what you think.

**Comment by potiuk** [comment]
Closing for now as we perform cleanup and already changed the implementation/formatting quite a bit. We might reopen it as needed and will use it in case we come back to this issue.
